### DISCUSSION TITLE
Optimized AES Memory Footprint

### DISCUSSION THREAD

**Discussion by pa-pa** [pa-pa/AskSinPP#278] [Show and tell]
Because of some problems with my pool control I assume, that the stack space is not big enough. I reviewed the AES implementation and removed some of the static object and used stack objects. This reduces the footprint of all devices a lot, when using AES.
In addition I reduced the size of a message to the current known minimum - 17 byte payload.
Before I merge this into the master I would like to know, if there are problems with this changes. So if you use the AES functionality please verify that the code is working after switching to this branch.

**Comment by stan23** [comment]
#define MaxDataLen 17

This is only valid for HM messages, but not for HmIP, where it can be even more than 30 byte:

https://youtu.be/uAyzimU60jw?t=581
http://www.homematic-inside.de/media/download/legacy/meetup_2018/sendeverhalten
slides 4 and 9

For sure AES affects only HM messages, but I see the risk to run into another buffer overflow if a longer packet is received.

Btw, my fix and adaption for the short HmIP messages seems to work smoothly for my problematic dimmer device:
https://github.com/pa-pa/AskSinPP/compare/master...stan23:dev-msgLength?expand=1

**Reply by pa-pa** [reply]
We are check if the received package fits into the provided buffer.

https://github.com/pa-pa/AskSinPP/blob/711842f791a5365f8a928596fca4ab7af4155dd8/Radio.h#L776

So this should OK for HM Classic stuff.

**Reply by stan23** [reply]
Okay then it is fine regarding the smaller buffer size.

**Reply by stan23** [reply]
But it would affect the AskSinAnalyzer if HmIP messages with longer payload are simply dropped.

Maybe there should be a #define for it?
"Normal" AskSin devices do not care about HmIP messages at all and don't mind if the long messages are dropped.
AskSinAnalyzer will analyse and show also the long messages:

```
#ifdef RECEIVE_LONG_HMIP_MESSAGES
#define MaxDataLen 51
#else
#define MaxDataLen 17
#endif
```

**Comment by pa-pa** [comment]
Hab das jetzt mal einfach so eingebaut:

#ifndef MaxDataLen
  //#define MaxDataLen   60						// maximum length of received bytes
  #define MaxDataLen   17
#endif
